multicluster/service-mirror/cluster_watcher.go

```diff
@@ -324,8 +339,34 @@ func (rcsw *RemoteClusterServiceWatcher) cleanupMirroredResources(ctx context.Co
 // Deletes a locally mirrored service as it is not present on the remote cluster anymore
 func (rcsw *RemoteClusterServiceWatcher) handleRemoteServiceDeleted(ctx context.Context, ev *RemoteServiceDeleted) error {
 	localServiceName := rcsw.mirroredResourceName(ev.Name)
-	rcsw.log.Infof("Deleting mirrored service %s/%s", ev.Namespace, localServiceName)
+	localService, err := rcsw.localAPIClient.Svc().Lister().Services(ev.Namespace).Get(localServiceName)
 	var errors []error
+	if err != nil {
+		errors = append(errors, fmt.Errorf("could not fetch Service %s/%s: %s", ev.Namespace, localServiceName, err))
+	}
+
+	// If the mirror service is headless, also delete its endpoint mirror
+	// services.
+	if rcsw.headlessServicesEnabled && localService.Spec.ClusterIP == corev1.ClusterIPNone {
+		matchLabels := map[string]string{
+			consts.MirroredHeadlessSvcNameLabel: localServiceName,
+		}
+		endpointMirrorServices, err := rcsw.localAPIClient.Svc().Lister().List(labels.Set(matchLabels).AsSelector())
+		if err != nil {
+			errors = append(errors, fmt.Errorf("could not fetch Endpoint Mirrors for Service %s/%s: %s", ev.Namespace, localServiceName, err))
+		}
+
+		for _, endpointMirror := range endpointMirrorServices {
+			err = rcsw.localAPIClient.Client.CoreV1().Services(endpointMirror.Namespace).Delete(ctx, endpointMirror.Name, metav1.DeleteOptions{})
```

Comment:
Do we also need to delete the Endpoints of each endpoint mirror service?  Or does kubernetes automatically delete them when the associated service is deleted?

Reply:
Good question. The endpoints should be automatically deleted by k8s when the associated service is deleted.